storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1268 @@
+	uint32_t rca;
+	/* Size of SD card in blocks */
+	uint32_t sizeBlocks;
+	/* Size (in blocks) of an erase sector */
+	uint32_t eraseSizeBlocks;
+	/* Number of command timeouts since last card initialization */
+	uint32_t commandTimeouts;
+	/* Whether the card supports SDHC protocol (requires slightly different handling) */
+	uint8_t highCapacity;
+} sdcard_cardMetadata_t;
+
+typedef struct {
+	/* Base address (in virtual memory) of SD Host Controller register set */
+	volatile uint32_t *base;
+	handle_t cmdLock;
+	sdcard_cardMetadata_t card;
+	/* Reference clock frequency in Hz */
+	uint32_t refclkFrequency;
+
+	/* Address of DMA buffer in virtual memory (for access by the CPU) */
+	void *dmaBuffer;
+	/* Address of DMA buffer in physical memory (for access by the SD Host Controller) */
+	addr_t dmaBufferPhys;
+
+	uint8_t sdioInitialized;
+	uint8_t isCDPinSupported;
+	uint8_t isWPPinSupported;
+
+	handle_t isrHandle;
+	handle_t eventLock;
+	handle_t eventCond;
+} sdcard_hostData_t;
+
+static sdcard_hostData_t sdio_hosts[PLATFORM_SDIO_N_HOSTS] = { 0 };
+static unsigned int initializedHosts = 0;
+
+static struct {
+	handle_t lock;
+	handle_t cond;
+	uint8_t initialized;
+} presenceEvents = { .initialized = 0 };
+
+static inline sdcard_hostData_t *sdcard_getHostForSlot(unsigned int slot)
+{
+	return slot >= PLATFORM_SDIO_N_HOSTS ? NULL : &sdio_hosts[slot];
+}
+
+#define GET_HOST_OR_RETURN(...) \
+	sdcard_hostData_t *host = sdcard_getHostForSlot(slot); \
+	if (host == NULL) { \
+		return __VA_ARGS__; \
+	}
+
+static int sdcard_configClockAndPower(sdcard_hostData_t *host, uint32_t freq);
+static int sdcard_wideAndFast(sdcard_hostData_t *host);
+
+/* Resets parts or all of the host according to the given reset type.
+ * Argument must be one of CLOCK_CONTROL_RESET_*
+ */
+static int sdhost_reset(sdcard_hostData_t *host, uint32_t resetType)
+{
+	uint8_t i;
+	uint32_t val;
+
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
+	for (i = 0; i < SDHOST_RETRIES; ++i) {
+		usleep(10);
+		val = *(host->base + SDHOST_REG_CLOCK_CONTROL) & resetType;
+		if (val == 0) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+
+static inline uint8_t sdcard_isWriteProtected(sdcard_hostData_t *host)
+{
+	int presState;
+	if (host->isWPPinSupported) {
+		presState = *(host->base + SDHOST_REG_PRES_STATE);
+		if ((presState & PRES_STATE_WRITE_PROT_PIN) == 0) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+
+static void *sdhost_mmapDMA(size_t size)
+{
+	void *p;
+	size_t sz = (size + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
+
+	p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, NULL, 0);
+	if (p == MAP_FAILED) {
+		return NULL;
+	}
+
+	return p;
+}
+
+
+/* Checks physical address of a buffer in virtual memory and checks if the buffer is contiguous. */
+static addr_t sdhost_mphys(void *v, size_t *vSize)
+{
+	size_t sz;
+	addr_t phys;
+	addr_t nextPhys;
+
+	phys = va2pa(v);
+	sz = _PAGE_SIZE - (phys & (_PAGE_SIZE - 1));
```

Comment:
We decided to limit variables scope
```suggestion
	addr_t nextPhys;

	addr_t phys = va2pa(v);
	size_t sz = _PAGE_SIZE - (phys & (_PAGE_SIZE - 1));
```